accord-core/src/main/java/accord/impl/AbstractFetchCoordinator.java

```
@@ -0,0 +1,288 @@
+            {
+                if (!reply.isOk())
+                {
+                    fail(to, new RuntimeException(reply.toString()));
+                    inflight.remove(key).cancel();
+                    switch ((ReadData.ReadNack) reply)
+                    {
+                        default: throw new AssertionError("Unhandled enum: " + reply);
+                        case Invalid:
+                        case Redundant:
+                        case NotCommitted:
+                            throw new AssertionError(String.format("Unexpected reply: %s", reply));
+                        case Error:
+                            // TODO (required): ensure errors are propagated to coordinators and can be logged
+                    }
+                    return;
+                }
+
+                FetchResponse ok = (FetchResponse) reply;
+                Ranges received;
+                if (ok.unavailable != null)
+                {
+                    unavailable(to, ok.unavailable);
+                    if (ok.data == null)
+                    {
+                        inflight.remove(key).cancel();
+                        return;
+                    }
+                    received = ranges.difference(ok.unavailable);
+                }
+                else
+                {
+                    received = ranges;
+                }
+
+                // TODO (now): make sure it works if invoked in either order
+                inflight.remove(key).started(ok.maxApplied);
+                onReadOk(to, commandStore, ok.data, received);
+                // received must be invoked after submitting the persistence future, as it triggers onDone
+                // which creates a ReducingFuture over {@code persisting}
+            }
+
+            @Override
+            public void onFailure(Node.Id from, Throwable failure)
+            {
+                inflight.remove(key).cancel();
```

it would be nice if `cancel` could take the failure... so we could have a global handling of the throwable rather than the caller of this coordinator being the only one that could

---

what would be the benefit there?

---

> what would be the benefit there?

really its just knowing why we failed rather than going silent...  This is a network failure most likely so its really more likely to be `UNKNOWN`, which isn't 100% helpful... but yeah (unknown means "go check the FROM", but if its a network related issue, it might be safe to just bounce and try again?), I just tend to get squeamish when I see failures not checked